consensus/istanbul/backend/announce.go

```diff
@@ -558,161 +619,421 @@ func (sb *Backend) validateAnnounce(announceData *announceData) (bool, error) {
+// not being regossiped due to the cooldown period enforced by the other nodes in
+// the network. This is somewhat circumvented by caching the hashes of messages that
+// are regossiped to prevent future regossips, but this isn't guaranteed fails to
+// account for situations when an announcing node expects a message to be
+// gossiped out but gets rejected due to a cooldown period.
+// To ensure that a new legitimate message will eventually be gossiped to the rest
+// of the network, we schedule an announce message with a new timestamp that is
+// received during the cooldown period to be sent after the cooldown period ends,
+// and refuse to regossip older messages from the validator during that time.
+// Providing `force` as true will skip any of the DOS checks.
+func (sb *Backend) regossipAnnounce(msg *istanbul.Message, msgTimestamp uint, payload []byte, force bool) error {
+	logger := sb.logger.New("func", "regossipAnnounce", "announceSourceAddress", msg.Address, "msgTimestamp", msgTimestamp)
+
+	if !force {
+		sb.scheduledAnnounceRegossipsMu.RLock()
+		scheduledRegossip := sb.scheduledAnnounceRegossips[msg.Address]
+		// if a regossip for this msg address has been scheduled already, override
+		// the scheduling if this Timestamp is newer. Otherwise, don't regossip this message
+		// and let the scheduled regossip occur.
+		if scheduledRegossip != nil {
+			if msgTimestamp > scheduledRegossip.MsgTimestamp {
+				sb.scheduledAnnounceRegossipsMu.RUnlock()
+				return sb.scheduleAnnounceRegossip(msg, msgTimestamp, payload)
+			}
+			sb.scheduledAnnounceRegossipsMu.RUnlock()
+			logger.Debug("Announce message is not greater than scheduled regossip version, not regossiping")
 			return nil
 		}
+		sb.scheduledAnnounceRegossipsMu.RUnlock()
+
+		sb.lastAnnounceGossipedMu.RLock()
+		if lastGossiped, ok := sb.lastAnnounceGossiped[msg.Address]; ok {
+			if time.Since(lastGossiped.Time) < announceGossipCooldownDuration {
+				// If this timestamp is newer than the previously regossiped one,
+				// schedule it to be regossiped once the cooldown period is over
+				if lastGossiped.MsgTimestamp < msgTimestamp {
+					sb.lastAnnounceGossipedMu.RUnlock()
+					logger.Trace("Already regossiped msg from this source address with an older announce version within the cooldown period, scheduling regossip for after the cooldown")
+					return sb.scheduleAnnounceRegossip(msg, msgTimestamp, payload)
+				}
+				sb.lastAnnounceGossipedMu.RUnlock()
+				logger.Trace("Already regossiped msg from this source address within the cooldown period, not regossiping.")
+				return nil
+			}
+		}
+		sb.lastAnnounceGossipedMu.RUnlock()
 	}
-	sb.lastAnnounceGossipedMu.RUnlock()
 
 	logger.Trace("Regossiping the istanbul announce message", "IstanbulMsg", msg.String())
 	if err := sb.Multicast(nil, payload, istanbulAnnounceMsg); err != nil {
 		return err
 	}
 
-	sb.lastAnnounceGossipedMu.Lock()
-	defer sb.lastAnnounceGossipedMu.Unlock()
```

Comment:
I seems like we still need this lock. Writing to this array from multiple threads can cause a panic.